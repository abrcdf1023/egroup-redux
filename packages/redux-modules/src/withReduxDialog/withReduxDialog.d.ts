import * as React from 'react';

interface ReduxDialogProps {
  initializeDialog: Function;
  closeDialog: Function;
}

declare function withReduxDialog(
  dialogName: string
): React.ComponentType<ReduxDialogProps>;

export default withReduxDialog;
