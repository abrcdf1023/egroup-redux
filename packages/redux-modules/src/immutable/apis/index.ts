export * from './apis';
export * from '../../apis/types';
export * from '../../apis/actions';
