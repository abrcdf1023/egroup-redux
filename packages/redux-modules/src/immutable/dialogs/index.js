export * from './dialogs';
export * from './selectors';
export * from '../../dialogs/types';
export * from '../../dialogs/actions';
