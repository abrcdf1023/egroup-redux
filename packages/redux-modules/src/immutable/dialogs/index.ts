export { default as withReduxDialog } from './withReduxDialog';
export * from './withReduxDialog';
export * from './dialogs';
export * from './selectors';
export * from '../../dialogs/types';
export * from '../../dialogs/actions';
