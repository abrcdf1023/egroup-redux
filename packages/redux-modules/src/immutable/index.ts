export * from './apis';
export * from './dialogs';
export * from './entities';
export * from './snackbars';
