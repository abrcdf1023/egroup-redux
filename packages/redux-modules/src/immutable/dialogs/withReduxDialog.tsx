import React, { ComponentType, Component, RefObject, forwardRef } from 'react';
import { connect } from 'react-redux';

import { initializeDialog, closeDialog } from '../../dialogs/actions';
import { getDialogStates } from './selectors';

export interface OwnProps {
  handleClose: () => void;
}

export interface DispatchProps {
  initializeDialog: (name: string) => void;
  closeDialog: (name: string) => void;
}

export interface StateProps {
  isOpen: boolean;
}

/**
 * Please read this article for more info.
 * https://medium.com/@martin_hotell/react-refs-with-typescript-a32d56c4d315
 * @param name
 */
const withReduxDialog = (name: string) => <
  T extends Component,
  OriginalProps extends {}
>(
  WrappedComponent: ComponentType<any | string>
) => {
  type PrivateProps = { forwardedRef: RefObject<T> };

  type Props = StateProps & DispatchProps & OwnProps & PrivateProps;

  class WithReduxDialog extends Component<Props> {
    componentDidMount() {
      this.props.initializeDialog(name);
    }
    render() {
      const {
        forwardedRef,
        initializeDialog,
        handleClose,
        closeDialog,
        ...other
      } = this.props;
      return (
        <WrappedComponent
          ref={forwardedRef}
          handleClose={() => {
            closeDialog(name);
          }}
          {...other}
        />
      );
    }
  }

  /**
   * Connect before forwardRef
   * https://github.com/reduxjs/react-redux/issues/914
   */
  const mapStateToProps = (state: any, ownProps: OwnProps): StateProps => ({
    ...getDialogStates(state, ownProps, name).toJS()
  });

  const mapDispatchToProps = (): DispatchProps => ({
    initializeDialog,
    closeDialog
  });

  const ConnectedComponent = connect<StateProps, DispatchProps, OwnProps>(
    mapStateToProps,
    mapDispatchToProps
  )(WithReduxDialog);

  /**
   * Forwarding refs in higher-order components
   * https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components
   */
  function RefForwardingFactory(props: any, ref: T) {
    return <ConnectedComponent {...props} forwardedRef={ref} />;
  }

  // Give this component a more helpful display name in DevTools.
  const componentName = WrappedComponent.displayName || WrappedComponent.name;
  RefForwardingFactory.displayName = `withReduxDialog(${componentName})`;

  return forwardRef<T, OriginalProps>(RefForwardingFactory as any);
};

export default withReduxDialog;
