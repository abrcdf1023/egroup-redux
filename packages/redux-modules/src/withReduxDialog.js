import React, { Component } from 'react';
import { connect } from 'react-redux';

import { getRootStates, initializeDialog, closeDialog } from './dialogs';

const withReduxDialog = name => Dialog => {
  class ReduxDialog extends Component {
    componentDidMount() {
      this.props.initializeDialog(name);
    }
    render() {
      const { initializeDialog, handleClose, ...other } = this.props;
      return (
        <Dialog
          handleClose={() => {
            this.props.closeDialog(name);
          }}
          {...other}
        />
      );
    }
  }

  const mapStateToProps = (state, props) => ({
    ...getRootStates(state, props, name).toJS()
  });

  return connect(
    mapStateToProps,
    {
      initializeDialog,
      closeDialog
    }
  )(ReduxDialog);
};

export default withReduxDialog;
