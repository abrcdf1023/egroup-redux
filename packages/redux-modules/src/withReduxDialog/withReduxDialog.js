import React from 'react';
import { connect } from 'react-redux';

import { getDialogStates, initializeDialog, closeDialog } from '../dialogs';

const withReduxDialog = name => Dialog => {
  class ReduxDialog extends React.Component {
    componentDidMount() {
      this.props.initializeDialog(name);
    }
    render() {
      const {
        forwardedRef,
        initializeDialog,
        handleClose,
        closeDialog,
        ...other
      } = this.props;
      return (
        <Dialog
          ref={forwardedRef}
          handleClose={() => {
            closeDialog(name);
          }}
          {...other}
        />
      );
    }
  }

  /**
   * Connect before forwardRef
   * https://github.com/reduxjs/react-redux/issues/914
   */
  const mapStateToProps = (state, props) => ({
    ...getDialogStates(state, props, name)
  });

  const ConnectedComponent = connect(mapStateToProps, {
    initializeDialog,
    closeDialog
  })(ReduxDialog);

  /**
   * Forwarding refs in higher-order components
   * https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components
   */
  function forwardRef(props, ref) {
    return <ConnectedComponent {...props} forwardedRef={ref} />;
  }

  // Give this component a more helpful display name in DevTools.
  const dialogComponentName = Dialog.displayName || Dialog.name;
  forwardRef.displayName = `withReduxDialog(${dialogComponentName})`;

  return React.forwardRef(forwardRef);
};

export default withReduxDialog;
