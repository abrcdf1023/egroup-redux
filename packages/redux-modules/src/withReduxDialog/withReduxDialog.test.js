import '@testing-library/jest-dom/extend-expect';

import React from 'react';
import { render } from '@testing-library/react';
import { fromJS } from 'immutable';
import configureStore from 'redux-mock-store';
import withReduxDialog from './withReduxDialog';

let store;
const dialogName = 'alertAialog';

const MockDialog = ({ isOpen, handleClose, title, message }) => {
  return (
    <div>
      <p>{title}</p>
      <p>{message}</p>
    </div>
  );
};

describe('Redux Dialog HOC', () => {
  beforeEach(() => {
    const mockStore = configureStore();

    store = mockStore(
      fromJS({
        dialogs: {
          [dialogName]: {
            isOpen: false,
            title: 'dialog title',
            message: 'dialog message'
          }
        }
      })
    );
  });
  it('Should render the component only when dialog prop is true', () => {
    const ReduxDialog = withReduxDialog(dialogName)(MockDialog);
    const { getByText } = render(<ReduxDialog store={store} />);
    expect(getByText('dialog title')).toBeInTheDocument();
    expect(getByText('dialog message')).toBeInTheDocument();
  });
});
