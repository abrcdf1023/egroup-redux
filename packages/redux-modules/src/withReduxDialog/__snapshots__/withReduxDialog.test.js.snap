// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Dialog HOC Should render the component only when dialog prop is true 1`] = `
<div>
  <div>
    <p>
      dialog title
    </p>
    <p>
      dialog message
    </p>
  </div>
</div>
`;
